Add appVersion to the generated Chart.yaml file

### Component

Kubernetes Maven Plugin

### Is your enhancement related to a problem? Please describe

I would like to get the appVersion added to the chart, so that when I do

```bash
 helm list -n dev
```

When I do a simple:

```
./mvnw -pl microservice-app k8s:resource k8s:helm
```

The generated `Chart.yaml` contains a `version` key, but no `appVersion` key

```yaml
apiVersion: v1
name: microservice-app
<snip>
version: "0.3.1"
<snip>
```



Output is 
```
NAME            	NAMESPACE	REVISION	UPDATED                             	STATUS  	CHART                 	APP VERSION
microservice-app	dev      	1       	2023-11-23 10:58:51.765527 +0100 CET	deployed	microservice-app-0.3.1	0.3.1
```

Rather than

```
NAME            	NAMESPACE	REVISION	UPDATED                             	STATUS  	CHART                          	APP VERSION
microservice-app	dev      	1       	2023-11-23 10:32:28.876677 +0100 CET	deployed	microservice-app-0.3.1

```

This was mentioned in the following issue 3 years ago, at the time when helm 2 apparently didn't have `appVersion` https://github.com/eclipse/jkube/issues/113 


### Describe the solution you'd like

I would like the `appVersion` to be automatically added with the project version (same behaviour as `version`) 

### Describe alternatives you've considered

Another option is to have Chart resource fragments that allow adding a `Chart.helm.yml` in `src/main/jkube`.
And if this file contains
```yaml
appVersion: ${project.version}
```
then this gets templated and added to the output `Chart.yml`

### Additional context

_No response_

@manusa thanks a lot!

please note however that if you add `appVersion` as is, then `helm lint` will complain.

```bash
./mvnw clean install -Dmaven.test.skip
./mvnw -pl microservice-app k8s:resource k8s:helm
helm lint microservice-app/target/jkube/helm/microservice-app/kubernetes/microservice-app-0.0.1-SNAPSHOT.tar.gz
```

```
==> Linting microservice-app/target/jkube/helm/microservice-app/kubernetes/microservice-app-0.0.1-SNAPSHOT.tar.gz
[ERROR] Chart.yaml: chart type is not valid in apiVersion 'v1'. It is valid in apiVersion 'v2'

Error: 1 chart(s) linted, 1 chart(s) failed
```

Using fragments, I overrode the `apiVersion` with an entry in `Chart.helm.yaml`.

I think that if you want to add it unconditionnally, you would need to bump `apiVersion`

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -26,6 +26,7 @@ Usage:
 * Fix #2390: support for all missing Chart.yaml fields
 * Fix #2444: Add support for Spring Boot application properties placeholders
 * Fix #2456: Add utility class to decompress archive files
+* Fix #2472: Support for Helm Chart.yaml appVersion field defaulting to project version
 
 ### 1.15.0 (2023-11-10)
 * Fix #2138: Support for Spring Boot Native Image

--- a/gradle-plugin/it/src/it/helm-dsl/expected/Chart.yaml
+++ b/gradle-plugin/it/src/it/helm-dsl/expected/Chart.yaml
@@ -13,6 +13,7 @@ maintainers:
     email: maintainer@example.com
     url: https://example.com/user1
 icon: https://example.com/icon1
+appVersion: 0.0.1-SNAPSHOT
 dependencies:
   - name: dependency-via-groovy-dsl-config
     version: 0.0.1

--- a/gradle-plugin/it/src/it/helm-fragment-and-dsl/expected/Chart.yaml
+++ b/gradle-plugin/it/src/it/helm-fragment-and-dsl/expected/Chart.yaml
@@ -13,6 +13,7 @@ maintainers:
   email: maintainer@example.com
   url: https://example.com/user1
 icon: https://example.com/icon-from-fragment-overrides-dsl
+appVersion: 0.0.1-SNAPSHOT
 dependencies:
 - name: dependency-via-groovy-dsl-config
   version: 0.0.1

--- a/gradle-plugin/it/src/it/helm-fragment/expected/Chart.yaml
+++ b/gradle-plugin/it/src/it/helm-fragment/expected/Chart.yaml
@@ -14,6 +14,7 @@ maintainers:
   email: maintainer@example.com
   url: https://example.com/user1
 icon: https://example.com/icon1
+appVersion: 0.0.1-SNAPSHOT
 dependencies:
 - name: dependency-via-fragment
   version: 0.0.1

--- a/gradle-plugin/it/src/it/helm-properties/expected/Chart.yaml
+++ b/gradle-plugin/it/src/it/helm-properties/expected/Chart.yaml
@@ -5,3 +5,4 @@ home: https://example.com
 version: 0.0.1-SNAPSHOT-from-properties
 description: Helm chart generation via properties integration test
 icon: https://example.com/chart1
+appVersion: 0.0.1-SNAPSHOT

--- a/gradle-plugin/it/src/it/helm-zero-config/expected/Chart.yaml
+++ b/gradle-plugin/it/src/it/helm-zero-config/expected/Chart.yaml
@@ -2,3 +2,4 @@
 apiVersion: v1
 name: helm-zero-config
 version: 0.0.1-SNAPSHOT
+appVersion: 0.0.1-SNAPSHOT

--- a/jkube-kit/doc/src/main/asciidoc/inc/helm/_jkube_helm.adoc
+++ b/jkube-kit/doc/src/main/asciidoc/inc/helm/_jkube_helm.adoc
@@ -101,9 +101,16 @@ endif::[]
 |
 
 | *icon*
-| The Chart URL to an SVG or PNG image to be used as an icon , default is extracted from the kubernetes manifest
+| The Chart URL to an SVG or PNG image to be used as an icon, default is extracted from the kubernetes manifest
 (`kubernetes.yml`) `jkube.eclipse.org/iconUrl` annotation if not provided.
 | `jkube.helm.icon`
+
+| *appVersion*
+| The version of the application that Chart contains
+ifeval::["{plugin-type}" == "maven"]
+, defaults to `${project.version}` if not provided.
+endif::[]
+| `jkube.helm.appVersion`
 
 | *keywords*
 | Comma separated list of keywords to add to the chart.

--- a/jkube-kit/resource/helm/src/main/java/org/eclipse/jkube/kit/resource/helm/HelmConfig.java
+++ b/jkube-kit/resource/helm/src/main/java/org/eclipse/jkube/kit/resource/helm/HelmConfig.java
@@ -14,7 +14,6 @@
 package org.eclipse.jkube.kit.resource.helm;
 
 import com.fasterxml.jackson.annotation.JsonProperty;
-import io.fabric8.openshift.api.model.Parameter;
 import io.fabric8.openshift.api.model.Template;
 import lombok.AllArgsConstructor;
 import lombok.Builder;
@@ -22,7 +21,6 @@ import lombok.EqualsAndHashCode;
 import lombok.Getter;
 import lombok.NoArgsConstructor;
 import lombok.Setter;
-import lombok.Singular;
 import org.apache.commons.lang3.StringUtils;
 import org.eclipse.jkube.kit.common.Maintainer;
 
@@ -54,6 +52,7 @@ public class HelmConfig {
   private List<String> sources;
   private List<Maintainer> maintainers;
   private String icon;
+  private String appVersion;
   private List<String> keywords;
   private String engine;
   private List<File> additionalFiles;

--- a/jkube-kit/resource/helm/src/main/java/org/eclipse/jkube/kit/resource/helm/HelmService.java
+++ b/jkube-kit/resource/helm/src/main/java/org/eclipse/jkube/kit/resource/helm/HelmService.java
@@ -248,6 +248,7 @@ public class HelmService {
       .sources(helmConfig.getSources())
       .maintainers(helmConfig.getMaintainers())
       .icon(helmConfig.getIcon())
+      .appVersion(helmConfig.getAppVersion())
       .keywords(helmConfig.getKeywords())
       .engine(helmConfig.getEngine())
       .dependencies(helmConfig.getDependencies())

--- a/jkube-kit/resource/helm/src/main/java/org/eclipse/jkube/kit/resource/helm/HelmServiceUtil.java
+++ b/jkube-kit/resource/helm/src/main/java/org/eclipse/jkube/kit/resource/helm/HelmServiceUtil.java
@@ -51,6 +51,7 @@ public class HelmServiceUtil {
   private static final String PROPERTY_API_VERSION = "jkube.helm.apiVersion";
   private static final String DEFAULT_API_VERSION = "v1";
   private static final String PROPERTY_ICON = "jkube.helm.icon";
+  private static final String PROPERTY_APP_VERSION = "jkube.helm.appVersion";
   private static final String PROPERTY_TYPE = "jkube.helm.type";
   private static final String PROPERTY_CHART = "jkube.helm.chart";
   private static final String PROPERTY_CHART_EXTENSION = "jkube.helm.chartExtension";
@@ -105,6 +106,7 @@ public class HelmServiceUtil {
       () -> String.format("%s/jkube/helm/%s", project.getBuildDirectory(), helmConfig.getChart())));
     helmConfig.setIcon(resolveFromPropertyOrDefault(PROPERTY_ICON, project, helmConfig::getIcon,
         () -> findIconURL(new File(helmConfig.getSourceDir()), helmConfig.getTypes())));
+    helmConfig.setAppVersion(resolveFromPropertyOrDefault(PROPERTY_APP_VERSION, project, helmConfig::getAppVersion, project::getVersion));
     helmConfig.setTarFileClassifier(resolveFromPropertyOrDefault(PROPERTY_TARBALL_CLASSIFIER, project, helmConfig::getTarFileClassifier, () -> EMPTY));
     helmConfig.setTarballOutputDir(resolveFromPropertyOrDefault(PROPERTY_TARBALL_OUTPUT_DIR, project, helmConfig::getTarballOutputDir,
         helmConfig::getOutputDir));

--- a/jkube-kit/resource/helm/src/test/java/org/eclipse/jkube/kit/resource/helm/HelmConfigTest.java
+++ b/jkube-kit/resource/helm/src/test/java/org/eclipse/jkube/kit/resource/helm/HelmConfigTest.java
@@ -102,6 +102,7 @@ class HelmConfigTest {
         "\"description\":\"The description\"," +
         "\"home\":\"e.t.\"," +
         "\"icon\":\"Warhol\"," +
+        "\"appVersion\":\"1.33.7\"," +
         "\"maintainers\":[{}]," +
         "\"sources\":[\"source\"]," +
         "\"engine\":\"V8\"," +
@@ -131,6 +132,7 @@ class HelmConfigTest {
         .hasFieldOrPropertyWithValue("description", "The description")
         .hasFieldOrPropertyWithValue("home", "e.t.")
         .hasFieldOrPropertyWithValue("icon", "Warhol")
+        .hasFieldOrPropertyWithValue("appVersion", "1.33.7")
         .hasFieldOrPropertyWithValue("maintainers", Collections.singletonList(new Maintainer()))
         .hasFieldOrPropertyWithValue("sources", Collections.singletonList("source"))
         .hasFieldOrPropertyWithValue("engine", "V8")
@@ -177,6 +179,7 @@ class HelmConfigTest {
     helmConfig.setDescription("description");
     helmConfig.setHome("home");
     helmConfig.setIcon("icon");
+    helmConfig.setAppVersion("1.33.7");
     helmConfig.setMaintainers(Collections.singletonList(new Maintainer()));
     helmConfig.setSources(Collections.singletonList("source"));
     helmConfig.setEngine("engine");
@@ -201,6 +204,7 @@ class HelmConfigTest {
     assertThat(helmConfig.getDescription()).isEqualTo("description");
     assertThat(helmConfig.getHome()).isEqualTo("home");
     assertThat(helmConfig.getIcon()).isEqualTo("icon");
+    assertThat(helmConfig.getAppVersion()).isEqualTo("1.33.7");
     assertThat(helmConfig.getMaintainers()).isNotEmpty();
     assertThat(helmConfig.getSources().get(0)).isEqualTo("source");
     assertThat(helmConfig.getEngine()).isEqualTo("engine");
@@ -226,6 +230,7 @@ class HelmConfigTest {
         .description("description")
         .home("e.t.")
         .icon("Warhol")
+        .appVersion("1.33.7")
         .maintainers(Collections.singletonList(new Maintainer()))
         .sources(Arrays.asList("source-1", "source-2"))
         .engine("V8")

--- a/jkube-kit/resource/helm/src/test/java/org/eclipse/jkube/kit/resource/helm/HelmServiceTest.java
+++ b/jkube-kit/resource/helm/src/test/java/org/eclipse/jkube/kit/resource/helm/HelmServiceTest.java
@@ -157,6 +157,7 @@ class HelmServiceTest {
       .keywords(Collections.singletonList("ci"))
       .maintainers(Collections.singletonList(Maintainer.builder().name("maintainer-from-config").build()))
       .icon("test-icon")
+      .appVersion("1.33.7")
       .engine("gotpl")
       .dependencies(Collections.singletonList(HelmDependency.builder().name("dependency-from-config").build()));
     // When
@@ -171,6 +172,7 @@ class HelmServiceTest {
       .hasFieldOrPropertyWithValue("description", "Description from helmconfig")
       .hasFieldOrPropertyWithValue("home", "https://example.com")
       .hasFieldOrPropertyWithValue("icon", "test-icon")
+      .hasFieldOrPropertyWithValue("appVersion", "1.33.7")
       .hasFieldOrPropertyWithValue("engine", "gotpl")
       .hasFieldOrPropertyWithValue("keywords", Collections.singletonList("fragment"))
       .hasFieldOrPropertyWithValue("sources", Collections.singletonList("https://source.example.com"))

--- a/kubernetes-maven-plugin/plugin/src/test/java/org/eclipse/jkube/maven/plugin/mojo/build/HelmMojoTest.java
+++ b/kubernetes-maven-plugin/plugin/src/test/java/org/eclipse/jkube/maven/plugin/mojo/build/HelmMojoTest.java
@@ -99,6 +99,7 @@ class HelmMojoTest {
       .hasFieldOrPropertyWithValue("description", "A description from Maven")
       .hasFieldOrPropertyWithValue("home", "https://project.url")
       .hasFieldOrPropertyWithValue("icon", null)
+      .hasFieldOrPropertyWithValue("appVersion", "1337")
       .hasFieldOrPropertyWithValue("sourceDir", projectDir.resolve("target")
         .resolve("classes").resolve("META-INF").resolve("jkube") + File.separator)
       .hasFieldOrPropertyWithValue("outputDir", projectDir.resolve("target")

